crates/generated_parser/src/early_error_checker.rs

```diff
@@ -0,0 +1,646 @@
+        match info.kind {
+            BindingKind::Let => {
+                context.declare_lex(
+                    info.name,
+                    DeclarationKind::Let,
+                    info.offset,
+                    &atoms.borrow(),
+                )?;
+            }
+            BindingKind::Const => {
+                context.declare_lex(
+                    info.name,
+                    DeclarationKind::Const,
+                    info.offset,
+                    &atoms.borrow(),
+                )?;
+            }
+            _ => {
+                panic!("Unexpected binding found {:?}", info);
+            }
+        }
+    }
+
+    Ok(())
+}
+
+pub trait EarlyErrorChecker<'alloc> {
+    fn context_metadata_mut(&mut self) -> &mut ContextMetadata;
+    fn context_metadata(&self) -> &ContextMetadata;
+    fn atoms(&self) -> &Rc<RefCell<SourceAtomSet<'alloc>>>;
+
+    // Check Early Error for BindingIdentifier and note binding info to the
+    // stack.
+    fn on_binding_identifier(&mut self, token: &arena::Box<'alloc, Token>) -> Result<'alloc, ()> {
+        let context = IdentifierEarlyErrorsContext::new();
+        context.check_binding_identifier(token, &self.atoms().borrow())?;
+
+        let name = token.value.as_atom();
+        let offset = token.loc.start;
+
+        if let Some(info) = self.context_metadata_mut().last_binding() {
+            debug_assert!(info.offset < offset);
+        }
+
+        self.context_metadata_mut().push_binding(BindingInfo {
+            name,
+            offset,
+            kind: BindingKind::Unknown,
+        });
+
+        Ok(())
+    }
+
+    // Check Early Error for IdentifierReference.
+    fn on_identifier_reference(&self, token: &arena::Box<'alloc, Token>) -> Result<'alloc, ()> {
+        let context = IdentifierEarlyErrorsContext::new();
+        context.check_identifier_reference(token, &self.atoms().borrow())
+    }
+
+    // Check Early Error for LabelIdentifier and note binding info to the
+    // stack
+    fn on_label_identifier(&mut self, token: &arena::Box<'alloc, Token>) -> Result<'alloc, ()> {
+        let context = IdentifierEarlyErrorsContext::new();
+
+        let name = token.value.as_atom();
+        let offset = token.loc.start;
+
+        if let Some(info) = self.context_metadata_mut().last_binding() {
+            debug_assert!(info.offset < offset);
+        }
+
+        // If the label is attached to a continue or break statement, its label info
+        // is popped from the stack. See `continue_statement` and `break_statement` for more
+        // information.
+        self.context_metadata_mut().push_label(LabelInfo {
+            name,
+            offset,
+            kind: LabelKind::Other,
+        });
+
+        context.check_label_identifier(token, &self.atoms().borrow())
+    }
+
+    fn check_labelled_statement(
+        &mut self,
+        name: SourceAtomSetIndex,
+        label_offset: usize,
```

Comment:
how about using `start_of_*_offset`, `end_of_*_offset` for all parameters in this file ?
that's used in some methods and I think it's clear where the offset should point.